zerver/lib/push_notifications.py

```diff
@@ -922,11 +922,74 @@ def process(elem: lxml.html.HtmlElement) -> str:
             plain_text += elem.tail or ""
         return plain_text
 
+    def is_same_server_message_link(hash: str) -> bool:
+        # A same server message link always has category `narrow`,
+        # section `stream` or `dm`, and ends with `/near/<message_id>`,
+        # where <message_id> is a sequence of digits.
+        match = re.match(r"#([^/]+)", hash)
+        if match is None or match.group(1) != "narrow":
+            return False
+
+        match = re.search(r"#narrow/([^/]+)", hash)
+        if match is None or not (match.group(1) == "stream" or match.group(1) == "dm"):
+            return False
+
+        return re.search(r"/near/\d+$", hash) is not None
```

Comment:
I feel like it would be better for this to be an algorithm that uses `urllib` for parsing the full link into search operand/operator pairs, and then doing checks on those operators.

We don't currently do this in the server, but it might make sense to have a new function in a new file `zerver/lib/url_decoding.py` for this logic, with its own little test suite...

Reply:
#30455